### PR TITLE
Fix batching partition semantics at week interval when the week span two years

### LINKED ISSUE #652: batch partition create week across the year partition name is wrong
### Steps to reproduce the behavior (Required)
1. CREATE TABLE test (
    k1 DATE,
    k2 INT,
    k3 SMALLINT,
    v1 VARCHAR(2048),
    v2 DATETIME DEFAULT "2014-02-04 15:36:00"
)
ENGINE=olap
DUPLICATE KEY(k1, k2, k3)
PARTITION BY RANGE (k1) (
    START ("2020-12-25") END ("2021-01-15") EVERY (INTERVAL 1 WEEK)
)
DISTRIBUTED BY HASH(k2) BUCKETS 10
PROPERTIES (
    "replication_num" = "1"
);
2. show create table test;

### Expected behavior (Required)
CREATE TABLE `test` (
  `k1` date NULL COMMENT "",
  `k2` int(11) NULL COMMENT "",
  `k3` smallint(6) NULL COMMENT "",
  `v1` varchar(2048) NULL COMMENT "",
  `v2` datetime NULL DEFAULT "2014-02-04 15:36:00" COMMENT ""
) ENGINE=OLAP 
DUPLICATE KEY(`k1`, `k2`, `k3`)
COMMENT "OLAP"
PARTITION BY RANGE(`k1`)
(PARTITION p2020_52 VALUES [('2020-12-25'), ('2021-01-01')),
PARTITION p2021_01 VALUES [('2021-01-01'), ('2021-01-08')),
PARTITION p2021_02 VALUES [('2021-01-08'), ('2021-01-15')))
DISTRIBUTED BY HASH(`k2`) BUCKETS 10 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
### Real behavior (Required)
CREATE TABLE `test` (
  `k1` date NULL COMMENT "",
  `k2` int(11) NULL COMMENT "",
  `k3` smallint(6) NULL COMMENT "",
  `v1` varchar(2048) NULL COMMENT "",
  `v2` datetime NULL DEFAULT "2014-02-04 15:36:00" COMMENT ""
) ENGINE=OLAP 
DUPLICATE KEY(`k1`, `k2`, `k3`)
COMMENT "OLAP"
PARTITION BY RANGE(`k1`)
(PARTITION p2020_52 VALUES [('2020-12-25'), ('2021-01-01')),
PARTITION p2021_53 VALUES [('2021-01-01'), ('2021-01-08')),
PARTITION p2021_01 VALUES [('2021-01-08'), ('2021-01-15')))
DISTRIBUTED BY HASH(`k2`) BUCKETS 10 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
1.19 before

### PR DESCRIPTION
**Pull request by Astralidea**
Fix #652 
JDK old API have this problem.
we could use new API to solve this problem.

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java [hunk 1]
```diff
@@ -127,6 +130,13 @@ public void setTimeUnit(String timeUnit) {
         List<SingleRangePartitionDesc> singleRangePartitionDescs = Lists.newArrayList();
         long currentLoopNum = 0;
         long maxAllowedLimit = Config.max_partitions_in_one_batch;
+
+        int dayOfWeek = 1;
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java:134]: 1. Please give some comments for this logical.
2. How about logic of adding partition.

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java [hunk 2]
```diff
@@ -136,7 +146,11 @@ public void setTimeUnit(String timeUnit) {
                     beginTime = beginTime.plusDays(timeInterval);
                     break;
                 case WEEK:
-                    partitionName = DEFAULT_PREFIX + beginTime.toString(DateUtils.WEEK_FORMAT);
+                    LocalDate localDate = LocalDate.of(beginTime.getYear(), beginTime.getMonthOfYear(),
+                            beginTime.getDayOfMonth());
+                    int weekOfWeekyear = localDate.get(weekFields.weekOfYear());
+                    partitionName = DEFAULT_PREFIX + String.format("%s_%02d",
+                            beginTime.toString(DateUtils.YEAR_FORMAT), weekOfWeekyear);
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java:153]: DEFAULT_PREFIX can be concat in String.format too.
That can reduce an string concat operation.

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java [hunk 3]
```diff
@@ -127,6 +131,16 @@ public void setTimeUnit(String timeUnit) {
         List<SingleRangePartitionDesc> singleRangePartitionDescs = Lists.newArrayList();
         long currentLoopNum = 0;
         long maxAllowedLimit = Config.max_partitions_in_one_batch;
+
+        // By default, the dayOfWeek parameter are the same as those of dynamic partition.
+        // If user set dynamic_partition.start_day_of_week table properties
+        // it will follow this configuration to set day of week
+        int dayOfWeek = 1;
+        if (properties != null && properties.get(DynamicPartitionProperty.START_DAY_OF_WEEK) != null) {
+            dayOfWeek = Integer.parseInt(properties.get(DynamicPartitionProperty.START_DAY_OF_WEEK));
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java:140]: When dayOfWeek is an invalid number, such as 9, can StarRocks handle it correctly?

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 4]
```diff
@@ -3078,11 +3078,13 @@ public void addPartitions(Database db, String tableName, AddPartitionClause addP
         } else if (partitionDesc instanceof MultiRangePartitionDesc) {
             db.readLock();
             RangePartitionInfo rangePartitionInfo;
+            Map<String, String> tableProperties;
             try {
                 Table table = db.getTable(tableName);
                 CatalogChecker.checkTableExist(db, tableName);
                 CatalogChecker.checkTableTypeOLAP(db, table);
                 OlapTable olapTable = (OlapTable) table;
+                tableProperties = olapTable.getTableProperty().getProperties();
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:3087]: Can olapTable.getTableProperty() return null?
can tableProperties be null?

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 5]
```diff
@@ -3100,8 +3102,13 @@ public void addPartitions(Database db, String tableName, AddPartitionClause addP
 
             Column firstPartitionColumn = partitionColumns.get(0);
             MultiRangePartitionDesc multiRangePartitionDesc = (MultiRangePartitionDesc) partitionDesc;
+            Map<String, String> properties = addPartitionClause.getProperties();
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:3105]: Is it possible that properties is null?

Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:3105]: Usually it won’t be null, but better add a check.

**Comment by chaoyli** [comment]
It's better to add concise commit message about the bug and solution.

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java [hunk 6]
```diff
@@ -127,6 +133,22 @@ public void setTimeUnit(String timeUnit) {
         List<SingleRangePartitionDesc> singleRangePartitionDescs = Lists.newArrayList();
         long currentLoopNum = 0;
         long maxAllowedLimit = Config.max_partitions_in_one_batch;
+
+        // By default, the dayOfWeek parameter are the same as those of dynamic partition.
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java:137]: 1. The problem is to make batching partition semantics is consistent with dynamic partition.
    In china, the Monday is the first day of week. In western country, the Sunday is the first day of week. The semantics is should be consistent between batching partition and dynamic partition. If the option is not set, the Monday will be the first day of week.
2. The 1st January is the first week of every year. Every year have 52 weeks. The last week will end at 31st December.

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java [hunk 7]
```diff
@@ -127,6 +133,22 @@ public void setTimeUnit(String timeUnit) {
...
+        // If user set dynamic_partition.start_day_of_week table properties
+        // it will follow this configuration to set day of week
+        int dayOfWeek = 1;
+        if (properties != null && properties.get(DynamicPartitionProperty.START_DAY_OF_WEEK) != null) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java:141]:
```suggestion
        if (properties != null && properties.containsKey(DynamicPartitionProperty.START_DAY_OF_WEEK)) {
```

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java (hunk 6 above)
Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java:137]: > The 1st January is the first week of every year. Every year have 52 weeks. The last week will end at 31st December.

+1

**Review by wyb — APPROVED** [review]
Please fix the bug of the natural week later

**Review by chaoyli — APPROVED** [review]

**Pull request merged**